### DISCUSSION TITLE
为什么在Windows上安装mingw版boost的时候跑的安装脚本是shell啊

### DISCUSSION THREAD

**Discussion by wanzzhehe** [xmake-io/xmake#3635] [Q&A] [answered]
<img width="1045" alt="image" src="https://user-images.githubusercontent.com/26498686/232263977-977fec82-a597-4918-b621-d31dfc4acd70.png">

xrepo安装boost在windows上是默认msvc，想切换到mingw的时候安装失败，查看上图的输出发现是试图运行 `./bootstrap.sh` 而不是 `bootstrap.bat`

**Comment by waruqi** [comment, marked as answer]
大部分 mingw 库编译，基本上都得依赖 msys 环境，尤其是 boost 这种大库，因为不光用了 mingw 还会去用 gnu make, autoconf, pkgconfig 以及各种 shell 脚本和命令，纯 cmd 很多是跑不过的。

boost 的 bootstrap.bat 通常是给 msvc 用的，不一定支持 mingw ，你可以试试，如果能过 可以提 pr 过来改进包

如果想少躺坑，就在 msys 下执行各种 mingw 包编译安装

**Reply by wanzzhehe** [reply]
根据您的回复，我理解了为什么要用msys；

然后我去看了 xmake-repo 下面 boost 的构建脚本，发现了一个问题，bootstrap.bat 和 bootstrap.sh 按照 boost 的脚本，它们接受的参数是不一样的，详细解释如下：

这个是 xmake-repo 下 boost 的 xmake.lua 构建脚本：

```lua
        if package:is_plat("windows") then
            import("core.tool.toolchain")
            local runenvs = toolchain.load("msvc"):runenvs()
            os.vrunv("bootstrap.bat", bootstrap_argv, {envs = runenvs})
        elseif package:is_plat("mingw") and is_host("windows") then
            os.vrunv("sh", table.join("./bootstrap.sh", bootstrap_argv))
            os.cp("./tools/build/src/engine/b2.exe", ".")
        else
            os.vrunv("./bootstrap.sh", bootstrap_argv)
        end
```

可以看到 bootstrap.bat 和 bootstrap.sh 都接受了相同的参数，但是观察下面 bootstrap.bat 中处理参数的部分，分别是第15行：

```bat
call .\build.bat %*
```

这一部分用于构建 b2.exe，然后是第33到第47行：

```bat
SET TOOLSET=msvc

IF "%1"=="gcc" SET TOOLSET=gcc
IF "%1"=="clang" SET TOOLSET=clang
IF "%1"=="borland" SET TOOLSET=embarcadero
IF "%1"=="vc71" SET TOOLSET=msvc : 7.1
IF "%1"=="vc8" SET TOOLSET=msvc : 8.0
IF "%1"=="vc9" SET TOOLSET=msvc : 9.0
IF "%1"=="vc10" SET TOOLSET=msvc : 10.0
IF "%1"=="vc11" SET TOOLSET=msvc : 11.0
IF "%1"=="vc12" SET TOOLSET=msvc : 12.0
IF "%1"=="vc14" SET TOOLSET=msvc : 14.0
IF "%1"=="vc141" SET TOOLSET=msvc : 14.1
IF "%1"=="vc142" SET TOOLSET=msvc : 14.2
IF "%1"=="vc143" SET TOOLSET=msvc : 14.3
```

而 build.bat 中处理参数的部分，也仅对 `%1` 也就是第一个参数进行了处理。

综上，bootstrap.bat 仅接受一个参数，也就是表明 toolset 的参数，通过控制 toolset 参数就可以实现不借助 msys 的情况下使用 mingw 工具链编译 boost 库。

**Reply by waruqi** [reply]
> 根据您的回复，我理解了为什么要用msys；
> 
> 然后我去看了 xmake-repo 下面 boost 的构建脚本，发现了一个问题，bootstrap.bat 和 bootstrap.sh 按照 boost 的脚本，它们接受的参数是不一样的，详细解释如下：
> 
> 这个是 xmake-repo 下 boost 的 xmake.lua 构建脚本：
> 
> ```lua
>         if package:is_plat("windows") then
>             import("core.tool.toolchain")
>             local runenvs = toolchain.load("msvc"):runenvs()
>             os.vrunv("bootstrap.bat", bootstrap_argv, {envs = runenvs})
>         elseif package:is_plat("mingw") and is_host("windows") then
>             os.vrunv("sh", table.join("./bootstrap.sh", bootstrap_argv))
>             os.cp("./tools/build/src/engine/b2.exe", ".")
>         else
>             os.vrunv("./bootstrap.sh", bootstrap_argv)
>         end
> ```
> 
> 可以看到 bootstrap.bat 和 bootstrap.sh 都接受了相同的参数，但是观察下面 bootstrap.bat 中处理参数的部分，分别是第15行：
> 
> ```batchfile
> call .\build.bat %*
> ```
> 
> 这一部分用于构建 b2.exe，然后是第33到第47行：
> 
> ```batchfile
> SET TOOLSET=msvc
> 
> IF "%1"=="gcc" SET TOOLSET=gcc
> IF "%1"=="clang" SET TOOLSET=clang
> IF "%1"=="borland" SET TOOLSET=embarcadero
> IF "%1"=="vc71" SET TOOLSET=msvc : 7.1
> IF "%1"=="vc8" SET TOOLSET=msvc : 8.0
> IF "%1"=="vc9" SET TOOLSET=msvc : 9.0
> IF "%1"=="vc10" SET TOOLSET=msvc : 10.0
> IF "%1"=="vc11" SET TOOLSET=msvc : 11.0
> IF "%1"=="vc12" SET TOOLSET=msvc : 12.0
> IF "%1"=="vc14" SET TOOLSET=msvc : 14.0
> IF "%1"=="vc141" SET TOOLSET=msvc : 14.1
> IF "%1"=="vc142" SET TOOLSET=msvc : 14.2
> IF "%1"=="vc143" SET TOOLSET=msvc : 14.3
> ```
> 
> 而 build.bat 中处理参数的部分，也仅对 `%1` 也就是第一个参数进行了处理。
> 
> 综上，bootstrap.bat 仅接受一个参数，也就是表明 toolset 的参数，通过控制 toolset 参数就可以实现不借助 msys 的情况下使用 mingw 工具链编译 boost 库。

那你可以试下，如果它只是用了mingw编译器的话，如果它还用了 pkgconfig 等其他工具去找依赖库，或者还用了 gnumake 等等，不依赖 msys 不可能能跑通，cmd 下找不到 pkgconfig 等工具

这就要看编译 boost 有没有用到其他工具，除了编译器。。因为很多 mingw 库编译，都要借助一些 msys 里面的 gnu 工具。。

**Reply by wanzzhehe** [reply]
我现在跑了一下，在使用 `bootstrap.bat gcc` 命令构建完成 b2.exe 后，使用 `b2.exe install --prefix=<PATH> --build-dir=build/ -d+2 -q -a toolset=gcc threading=multi` 命令的条件下能够完整构建成功整个 boost 库，添加 `link=static` 或者 `link=shared` 参数后也能正常构建。

**Reply by wanzzhehe** [reply]
在修改完 boost 的 xmake.lua 后，很神奇地发现脚本生成的 user-config.jam 会导致 boost 构建失败。

```
C:\Lib\boost_1_82_0>b2 -a -q -d+2 --build-dir=build --user-config=user-config.jam toolset=gcc threading=multi
error: at project-config.jam:6
error: duplicate initialization of gcc with the following parameters:
error: version = 12
error: previous initialization at C:\Lib\boost_1_82_0\user-config.jam:1
```

user-config.jam 的内容如下：

```
using gcc : : C:\\Lib\\mingw64\\bin\\x86_64-w64-mingw32-g++.exe ;
```

我现在想到的一个 workaround 的方法是将 boost 的 xmake.lua 脚本一部分改为：

```lua
        local argv =
        {
            "--prefix=" .. package:installdir(),
            "--libdir=" .. package:installdir("lib"),
            "-d2",
            "-j4",
            "--hash",
            "--layout=tagged-1.66", -- prevent -x64 suffix in case cmake can't find it
            -- "--user-config=user-config.jam",
            "-sNO_LZMA=1",
            "-sNO_ZSTD=1",
            "install",
            "threading=" .. (package:config("multi") and "multi" or "single"),
            "debug-symbols=" .. (package:debug() and "on" or "off"),
            "link=" .. (package:config("shared") and "shared" or "static")
        }

        if not is_plat("mingw") then
            table.insert(argv, "--user-config=user-config.jam")
        end
```

经过测试，这么修改后 boost 在本地成功构建。

**Reply by waruqi** [reply]
user-config.jam 还是需要的，不能删，要跟 xmake 使用的工具链对上，否则它自动找的 mingw gcc 不一定就是 xmake 使用的。。有可能会冲突。。

**Reply by waruqi** [reply]
> using gcc : : C:\\Lib\\mingw64\\bin\\x86_64-w64-mingw32-g++.exe ;

g++ 改成 gcc 试试。。

这个你自己调下吧。。

**Reply by waruqi** [reply]
如果完全可以脱离 .sh 。。你可以提个 pr 过来

**Reply by wanzzhehe** [reply]
> 这个你自己调下吧。。

我知道这个是怎么回事了。

在 bootstrap.bat 会生成 project-config.jam 文件，然后里面定义了 `using gcc`，然后就会与 user-config.jam 中的 `using gcc` 冲突。

这个估计我只能写个逻辑删除 project-config.jam 文件中的 `using gcc` 语句才能解决了。

bootstrap.bat 中相关代码如下：

```bat
ECHO.
ECHO Generating Boost.Build configuration in project-config.jam for %TOOLSET%...
ECHO # Boost.Build Configuration > project-config.jam
ECHO # Automatically generated by bootstrap.bat >> project-config.jam
ECHO. >> project-config.jam
ECHO import option ; >> project-config.jam
ECHO. >> project-config.jam
ECHO using %TOOLSET% ; >> project-config.jam
ECHO. >> project-config.jam
ECHO option.set keep-going : false ; >> project-config.jam
ECHO. >> project-config.jam
```

**Reply by waruqi** [reply]
io.replace patch 删下就好，或者走 add_patches

**Reply by wanzzhehe** [reply]
这个说实话可以去找boost去提BUG，因为使用msvc的话就没有这问题Orz

**Reply by waruqi** [reply]
> 这个说实话可以去找boost去提BUG，因为使用msvc的话就没有这问题Orz

提 bug 解决不了当前问题，等他们解决，不知道何年马月了，而且低版本还是用不了。。只能打 patch
